### DISCUSSION TITLE
What's in Token.tensor for non-transformer models (like Tok2Vec)?

### DISCUSSION THREAD

**Discussion by ldorigo** [explosion/spaCy#8898] [Help: Coding & Implementations] [answered]
Hi! 

See https://github.com/allenai/scispacy/issues/387 for more background on the question. 
Scispacy provides a model consisting of the following components:

```python
[('tok2vec', <spacy.pipeline.tok2vec.Tok2Vec at 0x7fe728e0fc20>),
 ('tagger', <spacy.pipeline.tagger.Tagger at 0x7fe843db99a0>),
 ('attribute_ruler', <spacy.pipeline.attributeruler.AttributeRuler at 0x7fe728d92f80>),
 ('lemmatizer', <spacy.lang.en.lemmatizer.EnglishLemmatizer at 0x7fe728dcb1c0>),
 ('parser', <spacy.pipeline.dep_parser.DependencyParser at 0x7fe729615940>),
 ('ner', <spacy.pipeline.ner.EntityRecognizer at 0x7fe729615be0>)]
```
We want to serialize large amounts of parsed text, so need to be space efficient, and we found that the majority of the space taken by serialized documents comes from the `doc.tensor` data.

I'm a bit confused as to what `doc.tensor` and `token.tensor` contain here? I originally thought it was static word (or token) embedding information, but that seems to live in `token.vector`. More confusingly, `token.tensor` seems to be context-dependent:

```python

sent_1 = nlp("I am happy")
sent_2 = nlp("I love sushi")

sent_1[0].tensor == sent_2[0].tensor # False
```
Why is that the case, since this pipeline doesn't contain transformers?

Thanks you for your help!

**Comment by Pandalei97** [comment]
Hello, the field `tensor` contains the contextualized embedding for the token or the doc. It is calculated by the tok2vec layer or the transformer layer. In a spacy pipeline, the tok2vec layer and the transformer layer have the same role, it's also for that reason that we can't combine tok2vec and transformer in a same model. If you have included your static vectors in the pipeline, you should be able to get the static vector by `token.vector`, otherwise, it will just take the same value as `.tensor`.

**Comment by adrianeboyd** [comment, marked as answer]
`doc.tensor` stores the context-dependent tensors from the last `tok2vec` component in the pipeline. If there are no vectors in the pipeline, then this data is used as a backoff for `doc/span/token.vector`. (Having this backoff behavior was not really a good decision, but we're stuck with it at this point.)

`doc.tensor` is used while the pipeline is running, but you probably don't need it afterwards. You can pass `exclude=["tensor"]` into `Doc.to_bytes()` to skip it while saving.

However, if you want to serialize a large number of docs efficiently, you should consider using `DocBin` instead of `Doc.to_bytes()`. It will be more efficient because it doesn't duplicate the strings and you can control which doc attributes are saved. See: https://spacy.io/usage/saving-loading#docs

**Reply by ldorigo** [reply]
Thanks for the answer. Could you point me to where I can read about what you mean by context-dependent tensors? My (admittedly limited) understanding of word2vec is that the vector for each word is static (as in, it is only modified during training)?  

Regarding the last part, we need random access to single docs, so `DocBin` is not an option - but I'm interested in understanding what you mean by "doesn't duplicate the strings"?

**Reply by adrianeboyd** [reply]
The static vectors from a tool like `word2vec` or `glove` are the same for each token, but we're using `tok2vec` more generically to mean anything that maps a token in context in a doc to a vector. It could be static or it could be context-dependent. Our default options like [`spacy.HashEmbedCNN.v2`](https://spacy.io/api/architectures#HashEmbedCNN) consider a window of surrounding tokens (see the note about `window_size` in the docs), so the resulting tensors are context-dependent.

You can try it out with `en_core_web_sm` to see. Try `the` in different contexts and look at `token.tensor` to see the values change in different sentences. The values are less useful for semantic comparisons than with static `word2vec` vectors because they've been trained for the particular tasks of tagging and parsing in this pipeline, so while you can run `token.similarity` with the tensors, we wouldn't really recommend it. See: https://spacy.io/usage/linguistic-features#vectors-similarity

`Doc.to_bytes` saves a list of strings for token texts and annotations with each doc, so every `doc` containing the word `the` will have an entry for `"the"` in it and an entry for `"DT"`, etc. With `DocBin`, there's only one `"the"` entry and one `"DT"` entry for all the docs. The repeated strings can add up a bit on disk, but if you need random access then `DocBin` isn't a good solution.

**Reply by mbrunecky** [reply]
Regarding "_Regarding the last part, we need random access to single docs, so DocBin is not an option_", where to you need that random access? I believe DocBin is meant for training data, where it saves space. When doing predictions I am always working with a single document.
